### ISSUE TITLE
[Feature Request] Assign Category to Journal

### ISSUE THREAD

**Issue by kevinmatthes** [AmmarAbouZor/tui-journal#9]
As I already mentioned in #6, I primarily use the application to take notes on the projects I develop.  Thus, I have multiple journals which belong to one project.  I would like to ask for adding a category feature such that I can assign one optional category to a journal which should also affect the sorting in the left-hand side bar.

Here are some more detailed information of what I assume would be helpful for the implementation:

- When a journal is created / edited (the journal's metadata), there might be a third line in the input window to enter a category name.
- Entering a category should be optional such that by default, a journal does not have a category.
- The category should be changeable just like the journal's title.
- The left-hand side bar should reflect the category assignments by sorting the journals also by their category.
- The category should have the highest precedence when sorting the journals in the left-hand side bar.
- The category should be shown in another colour in the left-hand side bar together with the other metadata of a journal.
- In the database, the category could be another column of type `TEXT` in the `entries` table.
- The default value of `category` in the database should be SQL's `NULL` (no data, yet).
- If the database does not already exist, the schema of `entries` should contain the `category` column when creating that table.
- If the database already exists, the schema of `entries` should be altered by adding the new column; this would allow for the loss-free migration of databases from previous versions of the application.
- For the JSON back-end, the category field should be made an `Option<String>` in the Serde definition due to not being mandatory.
- For the implementation in general, Rust's default `String` type would suffice (in contrast to the `content` field which also might be a Gap Buffer or Rope for performance benefits, instead).
- On release, this would be a minor in the terms of SemVer (new feature, no breaking changes).

I assume that this feature would be useful to many users as they probably already know a similar structuring technique from other note-taking applications they may have used before.  Thus, the following use cases are also covered by this feature (target audiences in parentheses).

- organising shopping lists by stores (general audience)
- sorting lecture notes by lecture (students)
- managing papers' citation metadata by research topic / project (researchers)
- structuring details on stories when writing a novel / developing a game (authors / game designers)

What is your opinion on this feature, @AmmarAbouZor?

**Comment by AmmarAbouZor** [comment]
Thanks for the suggestion the feature and the helpful details for the implementation. 
I find this feature very helpful and can move the app from something very simple to more advance management tool.
Here are my thoughts about the implementation:
- I would use tags instead of categories, and would allow each journal to have many tags
- Saving the tags in an array on backends which defaults to empty array would make the migrations of the existing journals very easy
- In edit/create journal popup I would create a third input box and the tags can be comma separated
- Inside the journal popup, I would add keybinding which opens the already used tags to choose of them
- in the journals left-hand list, I would add a function to filter the journals according to the tags (popup with tags open, and the user can select the wanted tags)
- The sorting function would be helpful but I would move to a different feature because we need to think about other sorting options (Date ascending, alphabetically etc...). Feel free to open another issues for it with more details about what would be helpful for you.

Here are my thoughts about this feature. @kevinmatthes what do you think about it?

In my plan for this app, I want to rethink some points about how the source-code is structure and make some enhancements for the rendering-loops + adding a spinner for async operation, and It would be helpful, to do these changes for simplicity before implementing new features, but sure I'll do #6 before starting with that since it's simple to add, but I think this feature could be postponed for a while (two weeks or something)

**Comment by kevinmatthes** [comment]
This roadmap sounds good! :+1:
